utils/watchdog.py

```diff
@@ -0,0 +1,155 @@
+from collections import deque
+import logging
+import os
+import signal
+import threading
+import time
+import traceback
+
+# Not available on Windows
+try:
+    import resource
+except ImportError:
+    resource = None
+
+try:
+    import psutil
+except ImportError:
+    psutil = None
+
+from utils.platform import Platform
+
+log = logging.getLogger(__name__)
+
+
+class Watchdog(object):
+    def destruct(self):
+        raise NotImplementedError('Subclasses must override')
+
+    def reset(self):
+        raise NotImplementedError('Subclasses must override')
+
+    def watch(self):
+        raise NotImplementedError('Subclasses must override')
+
+
+class WatchdogWindows(Watchdog, threading.Thread):
+    """ Simple watchdog for Windows (relies on psutil) """
+    def __init__(self, duration):
+        self._duration = int(duration)
+
+        threading.Thread.__init__(self)
+        self.tlock = threading.RLock()
+        self.reset()
+        self.start()
+
+    def destruct(self):
+        try:
+            log.error("Self-destructing...")
+            log.error(traceback.format_exc())
+        finally:
+            # This will kill the current process including the Watchdog's thread
+            psutil.Process().kill()
+
+    def reset(self):
+        log.debug("Resetting watchdog for %d" % self._duration)
+        with self.tlock:
+            self.expire_at = time.time() + self._duration
+
+    def watch(self):
+        while True:
+            if time.time() > self.expire_at:
+                self.destruct()
+            time.sleep(self._duration/20)
+
+    def run(self):
+        self.watch()
+
+
+class WatchdogPosix(object):
+    """
+    Simple signal-based watchdog. Restarts the process when:
+    * no reset was made for more than a specified duration
+    * (optional) a specified memory threshold is exceeded
+    * (optional) a suspicious high activity is detected, i.e. too many resets for a given timeframe.
+
+    **Warning**: Not thread-safe.
+    Can only be invoked once per process, so don't use with multiple threads.
+    If you instantiate more than one, you're also asking for trouble.
+    """
+    # Activity history timeframe
+    _RESTART_TIMEFRAME = 60
+
+    def __init__(self, duration, max_mem_mb=None, max_resets=None):
+        # Set the duration
+        self._duration = int(duration)
+        signal.signal(signal.SIGALRM, WatchdogPosix.self_destruct)
+
+        # Set memory usage threshold
+        if max_mem_mb is not None:
```

Comment:
Let's take the opportunity to kill the memory limit code. It was added as an experiment and it never really took off and is not documented anywhere